GDJS/GDJS/Extensions/Builtin/AsyncExtension.cpp

```diff
@@ -0,0 +1,58 @@
+/*
+ * GDevelop JS Platform
+ * Copyright 2008-2016 Florian Rival (Florian.Rival@gmail.com). All rights
+ * reserved. This project is released under the MIT License.
+ */
+#include "AsyncExtension.h"
+
+#include "GDCore/Events/Builtin/AsyncEvent.h"
+#include "GDCore/Events/CodeGeneration/EventsCodeGenerationContext.h"
+#include "GDCore/Events/CodeGeneration/EventsCodeGenerator.h"
+#include "GDCore/Extensions/Builtin/AllBuiltinExtensions.h"
+
+namespace gdjs {
+
+AsyncExtension::AsyncExtension() {
+  gd::BuiltinExtensionsImplementer::ImplementsAsyncExtension(*this);
+
+  GetAllEvents()["Async::Async"].SetCodeGenerator(
+      [](gd::BaseEvent &event_, gd::EventsCodeGenerator &codeGenerator,
+         gd::EventsCodeGenerationContext &parentContext) {
+        gd::AsyncEvent &event = dynamic_cast<gd::AsyncEvent &>(event_);
+
+        // Generate callback code
+        const auto callbackDescriptor = codeGenerator.GenerateCallback(
+            gd::String::From(codeGenerator.GenerateSingleUsageUniqueIdFor(
+                event.GetInstruction().GetOriginalInstruction().lock().get())),
+            parentContext, event.GetActions(),
+            event.HasSubEvents() ? &event.GetSubEvents() : nullptr);
+
+        // Generate code to backup the objects lists
+        gd::String objectsListsBackupCode =
+            parentContext.IsAsync()
+                ? "asyncObjectsList = "
```

Comment:
@arthuro555 This seems dangerous, you assign the parameter asyncObjectsList to a new version made from itself. What happens if this is called multiple times?
You will create an asyncObjectsList from asyncObjectsList, itself from asyncObjectsList.

Reply:
Async events guarantee that, after one, the current function ends, since the code of any actions/subevents afterward is put in a callback function after the asynchronous task has finished executing. So this piece of code cannot be twice in the same function. Agreed, ideally, we'd use a separate variable (`const newAsyncObjectsLists`) but that would add more complexity for generating the arguments of callback functions since they would not match up between the calling end and the invoked end. The definition of the variable is as an argument of the function, setting it only changes the value for the current function, which has reached its end anyway, making it fully harmless. Therefore I judged this as not important enough to be worth increasing the complexity of the arguments code generation.

Reply:
I was concerned about being in an async callback with two subevents having both two async actions.

I think this works because the two subevents will be generated in two different functions, but that's a detail that may not hold in the future.
I think it's safer to do `const parentAsyncObjectsLists = asyncObjectsLists; const asyncObjectsLists = gdjs.LongLivedObjectsList.from(parentAsyncObjectsList)`

Reply:
Actually the trick of declaring a parentXxx variable does not work well because the shadowing of variable does not work as expected ( `Cannot access 'asyncObjectsLists' before initialization`) :/ 
Might still be worth avoiding modifying the parameter so that we can in the future not be afraid of putting sub events in functions or not.

Reply:
That'd be a syntax error, one cannot declare a variable using the name of an argument: 

![image](https://user-images.githubusercontent.com/19349038/162333867-d35c597d-959c-481a-aa34-ba32c534c8d3.png)

The event code generation with one function per event is currently not just a detail in my opinion, but a pretty important implementation choice for GDJS. If we were to change it, though I don't see why we would do it in the foreseeable future for GDJS, I'm pretty sure other changes would be needed anyways. We could do such a change then, which would be trivial with scoped blocks and no arguments preventing the "overshadowing".